MWPW-161354: filter unused fields in the editor

yesil opened the pull request:
Resolve:
- https://jira.corp.adobe.com/browse/MWPW-161354
- https://jira.corp.adobe.com/browse/MWPW-162449
- https://jira.corp.adobe.com/browse/MWPW-166250

Test URLs:
- Before: https://main--mas--adobecom.hlx.live/studio.html
- After: https://mwpw-161354--mas--adobecom.hlx.live/studio.html?milolibs=stage--milo--adobecom

Axelcureno commented on studio/src/editors/merch-card-editor.js:
```diff
@@ -167,6 +171,28 @@ class MerchCardEditor extends LitElement {
+            const field = this.querySelector(`sp-field-group.toggle#${key}`);
+            if (field) field.style.display = 'block';
+        });
+        this.superWide = variant.size?.includes('super-wide');
```
might be trivial but some variants have other sizes like `wide`

yesil replied:
@Axelcureno wide is supported by default. this flag is to show hide the extra-wide option when supported by the variant.

yesil replied:
I also added flag for wide cards.

afmicka commented:
@yesil  seems like you are missing the latest (Mariia's fix is either missing or not workin on this branch https://github.com/adobecom/mas/pull/118). Could you please update?

afmicka commented:
@yesil Double-click on the CTA from the RTE opens OST on price offers instead of checkout.
<img width="1289" alt="Screenshot 2025-01-21 at 14 32 14" src="https://github.com/user-attachments/assets/ef4fdb08-cbff-4a5a-bcdd-d265495d19a5" />

afmicka commented:
@yesil Could there be more clear separation between the fields (either bigger gap or maybe keep just an outline line for the field)? This "description" and "Footer" titles are barely noticeable (looks like they belong to the previous fields).
<img width="441" alt="Screenshot 2025-01-21 at 14 59 57" src="https://github.com/user-attachments/assets/30bc5494-e734-468e-98f7-3297cc2b1a6b" />

afmicka reviewed (changes requested):
@yesil changes in any of the fields is not previewed anymore on the cards. Please check.

yesil commented on studio/src/editor-panel.js:
```diff
@@ -94,20 +92,12 @@ export default class EditorPanel extends LitElement {
-            this.disabled = false;
-        };
+        Store.fragments.discard.set(true);
+        await this.updateComplete;
```
this is causing render nothing / then render everything forcing fields to refresh their values from the fragment.

yesil commented on studio/src/store.js:
```diff
@@ -1,9 +1,9 @@
 import { Fragment } from './aem/fragment.js';
-import { getEditorPanel } from './editor-panel.js';
```
editor-panel already imports the store api.
transitive updates causing problems.

afmicka commented:
@yesil Duplicate and Publish fragment have the same issue that 'Save' had. Can't perform the action. 
![Screenshot 2025-01-27 at 14 23 13](https://github.com/user-attachments/assets/832d5d2d-85a5-4818-b725-bbd93f6817f7)

yesil commented on studio/src/editor-panel.js:
```diff
@@ -35,79 +35,51 @@ export default class EditorPanel extends LitElement {
     get fragmentStore() {
-        if (!this.fragmentStoreController.value) return null;
-        return this.fragmentStoreController.value;
+        if (this.discarded) return null;
```
forces merch-card-editor to not render so that the field states are reset.

afmicka commented:
@yesil what is this bracket after the tags? :) 
<img width="454" alt="Screenshot 2025-01-27 at 17 04 57" src="https://github.com/user-attachments/assets/f6d8f589-4b88-452c-b701-d46148d92a96" />

yesil commented:
@afmicka, oh, it is gone now.
![image](https://github.com/user-attachments/assets/fc43d676-e692-4c72-ad41-0a496e7f95c6)

afmicka commented:
@yesil Leaving here as a reminder. When you open a slice card, change the variant to suggested, and then back again to slice, the size dropdown field is empty. Though card has turned back to wide slice by default. 
<img width="1137" alt="Screenshot 2025-01-27 at 17 10 08" src="https://github.com/user-attachments/assets/ce1ca58a-e77c-44a7-8cf9-b61d1440814c" />

afmicka commented:
@yesil when you click clone on MAIN, it greys out the fields (fields and actions on top) indicating that the operations is happening (screenshot). On your branch, nothing happens in the editor, user is not sure if anything is going on. The fragment gets cloned at the end but is missing this visual experience.
<img width="725" alt="Screenshot 2025-01-29 at 11 35 59" src="https://github.com/user-attachments/assets/cb75545c-869d-4b4e-a6a2-bbc2f811bef4" />

yesil commented:
@afmicka  the placeholder behavior is valid w.rt. SWC. I don't know where to fix it. Otherwise I addressed all the issues you reported.
Added also a dialog for the deletion
<img width="516" alt="image" src="https://github.com/user-attachments/assets/8a6643a6-6daf-4123-bf62-f40479f2f9c8" />

afmicka commented:
@yesil I believe on main is also SWC, right? The changes are breaking Nala. The field value update is present on main at the moment. Lets discuss on a call either the fix or the test change.

3ch023 commented:
I can still edit the card while spinner is running.
Imo greyed out fields option was more clear, lets ask Nick&Vetri to chose

3ch023 commented:
@yesil discussed offline with @afmicka, we just need to change here
https://github.com/adobecom/mas/blob/830993e0d9f6c4011ce3c782b36f29743eedfaf9/nala/studio/studio.test.js#L154

instead of checking attribute, 
```await expect(
                await studio.editorPanel.locator(studio.editorTitle),
            ).toHaveAttribute('value', `${data.title}`);
```
we can check element value property and should work good
`await studio.editorPanel.locator(studio.editorTitle).value`

probably it makes sense to fix it in this PR

afmicka commented:
@yesil @3ch023 there is another issue with these changes. As a consequence of not updating the value, automation does not see the field changed in preview either. I can not reproduce manually but this is the image from github actions (and i get the same when i run test locally) - title is changed in the editor but not on the card
![test-failed-1](https://github.com/user-attachments/assets/8244f61a-1e63-4572-9d93-70db27b14a7b)

st-angelo-adobe commented on studio/src/store.js:
```diff
@@ -1,16 +1,37 @@
 const initialSearch = MasSearch.fromHash();
 const initialFilters = MasFilters.fromHash();
 
+function discardChanges() {
```
Personally I don't think this should be here. 
This file should be kept as simple as possible, with just stores.
This is editor logic, and should belong in the editor files imo. Otherwise this file will look very weird in a few months.

st-angelo-adobe commented on studio/src/editor-panel.js:
```diff
@@ -33,24 +31,44 @@ export default class EditorPanel extends LitElement {
-    createRenderRoot() {
-        return this;
-    }
+    inEdit = Store.fragments.inEdit;
```
This isn't how we're supposed to use stores. Especially since you're subscribing to them reactively below anyway, with a new controller `ReactiveController`, which to me seems redundant.
If you want to react to `inEdit`, you just use a `StoreController` for it. Same for `operation`. This looks like `StoreController` with extra steps 😄 

Again, we can talk tomorrow in the dev call, but most of the time we should just use `StoreControllers`. That is their purpose, to give you access to the store value, reactively, which is what you want in 95%+ cases.

yesil replied:
Actually I want to remove the controller logic except for the reactivity to store changes.
Currently there are two ways of doing the same thing.
All the data read/update should be done via the reactiveStores.
we don't really need more.
I could remove the controller totally it would still work.

yesil replied:
you will have to convince me why controllers are needed to access store data.

afmicka commented:
@yesil there is a regression with discard changes. When you make a change to the card and close without saving, the preview of the card stays in the changed state instead of getting back to the original state.

st-angelo-adobe commented on studio/src/studio.js:
```diff
@@ -60,6 +60,7 @@ class MasStudio extends LitElement {
 
     currentPage = new StoreController(this, Store.currentPage);
     commerceEnv = new StoreController(this, Store.commerceEnv);
+    currentPage = new StoreController(this, Store.currentPage);
```
duplicate assignment
